pkg/operator/environmentchecker/environmentcheckcontroller.go

```diff
@@ -0,0 +1,409 @@
+
+	klog.V(2).Infof("Scheduled the next check in %s", delay)
+	time.AfterFunc(delay, func() {
+		queue.Add(queueKey)
+	})
+
+	if c.shouldDegradeCluster(result) {
+		return result.CheckError
+	}
+
+	// if operand was not started previously and block upgrade is false and clusterdegrade is also false
+	// then and only then we should start CSI driver operator
+	if !c.operandControllerStarted && !result.BlockUpgrade && !result.ClusterDegrade {
+		go c.runConditionalController(ctx)
+		c.operandControllerStarted = true
+	}
+
+	// sync storage class
+	storageClassSyncResult := c.storageClassController.Sync(ctx, c.vSphereConnection)
+
+	if c.shouldDegradeCluster(storageClassSyncResult) {
+		return storageClassSyncResult.CheckError
+	}
+
+	storageClassUpdateConditionError := c.updateConditions(storageClassControllerName, storageClassSyncResult)
+
+	if storageClassUpdateConditionError != nil {
+		return storageClassUpdateConditionError
+	}
+
+	return c.updateConditions(c.name, result)
+}
+
+func (c *EnvironmentCheckController) shouldDegradeCluster(result checks.ClusterCheckResult) bool {
+	if result.ClusterDegrade {
+		return true
+	}
+
+	if result.BlockUpgrade {
+		// if we were supposed to only block upgrades but for some reason we have created CSIDriver and storageclass
+		// in past and future syncing fails then we should degrade the cluster
+		csiDriver, err := c.csiDriverLister.Get(utils.VSphereDriverName)
+
+		// if we can't fetch CSI driver lets not degrade the cluster, if cluster indeed needs degrading we will catch in
+		// next sync
+		if err != nil {
+			return false
+		}
+		annotations := csiDriver.GetAnnotations()
+		if _, ok := annotations[utils.OpenshiftCSIDriverAnnotationKey]; ok {
+			return true
+		}
+		sc, err := c.scLister.Get(storageClassName)
+		if err != nil {
+			return false
+		}
+		if sc != nil {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (c *EnvironmentCheckController) runConditionalController(ctx context.Context) {
+	var wg sync.WaitGroup
+	wg.Add(len(c.controllers))
+
+	for i := range c.controllers {
+		go func(index int) {
+			cc := c.controllers[index]
+			defer klog.Infof("%s controller terminated", cc.name)
+			defer wg.Done()
+			// if conditionController is not running and there were no errors we should run
+			// those controllers
+			cc.controller.Run(ctx, 1)
+		}(i)
+	}
+	wg.Wait()
+}
+
+func (c *EnvironmentCheckController) runClusterCheck(ctx context.Context,
+	infra *ocpv1.Infrastructure) (delay time.Duration, result checks.ClusterCheckResult, checkRan bool, immediateError error) {
+
+	checkerApiClient := &checks.KubeApiInterfaceImpl{
+		CSINodeLister:   c.csiNodeLister,
+		CSIDriverLister: c.csiDriverLister,
+		NodeLister:      c.nodeLister,
+	}
+
+	if c.vSphereConnection != nil {
+		secretConfigHash, err := c.getSecretConfigMapHash(ctx, infra)
+		if err != nil {
+			immediateError = err
+			return
+		}
+
+		if secretConfigHash == c.vSphereConnection.SecretConfigMapHash {
```

Comment:
cc @derek-pryor @jsafrane @bertinatto this implements a form of connection caching. But the implementation can be tidied up some more.

Reply:
If we want to fix the connection leak(s), we need a small and separate PR that we can backport to 4.9 too.

Reply:
I got rid of connection caching and went for logout at the end of `sync` function approach. I think later approach is superior to connection cache because in that case, we do not hold any connection to vcenter at all.

See attached screenshot - even though our operator does all of its things it holds 0 connections to vcenter. 
[](url)
![Screenshot from 2021-12-08 11-05-33](https://user-images.githubusercontent.com/278/145241890-1ede3c1e-5bc5-48f3-a013-b1fab3b63872.png)